File: classes/Intcode.php
<?php
class Intcode
{
    const STATUS_STOP    = 0;
    const STATUS_RUNNING = 1;
    const STATUS_WAIT    = 2;
    const STATUS_HALT    = 3;

    private $code;          // intcode
    private $status;        // 0: stop, 1: running, 2: wait, 3: halt
    private $signal;        // output signal
    private $pos;           // current opcode position
    private $inputs;        // input signals
    private $relative_base; // relative base

    public function __construct($intcode)
    {
        $this->code = explode(',', $intcode);
        $this->status = self::STATUS_RUNNING;
        $this->signal = false;
        $this->pos = 0;
        $this->inputs = [];
        $this->relative_base = 0;

        return true;
    }

    public function setCode($pos, $code)
    {
        $this->code[$pos] = $code;
        return true;
    }

    public function setInputs($inputs)
    {
        if (is_array($inputs)) {
            $this->inputs = array_merge($this->inputs, $inputs);
        } else {
            $this->inputs[] = $inputs;
        }

        $this->status = self::STATUS_RUNNING;

        return true;
    }

    private function getCodeValue($mode, $pos_index)
    {
        $pos_value = @$this->code[$this->pos + $pos_index];
        $return_value = false;

        switch (intval($mode)) {
            case 1:
                $return_value = $pos_value;
                break;
            case 2:
                $return_value = @$this->code[$this->relative_base + $pos_value];
                break;
            case 0:
            default:
                $return_value = @$this->code[$pos_value];
                break;
        }

        return $return_value;
    }

    private function setCodeValue($mode, $pos_index, $value)
    {
        $pos_value = @$this->code[$this->pos + $pos_index];
        switch (intval($mode)) {
            case 2:
                $this->code[$this->relative_base + $pos_value] = $value;
                break;
            case 0:
            case 1:
            default:
                $this->code[$pos_value] = $value;
                break;
        }
        return true;
    }

    public function calc()
    {
        $inst = sprintf('%05d', $this->code[$this->pos]);

        $op    = substr($inst, -2);
        $mode1 = substr($inst, 2, 1);
        $mode2 = substr($inst, 1, 1);
        $mode3 = substr($inst, 0, 1);

        $jump = false;

        switch ($op) {
            case '99':
                $this->status = self::STATUS_HALT;
                break;
            case '01':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];
                // $pos3 = $this->code[$this->pos + 3];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                $n3 = $n1 + $n2;
                // $this->code[$pos3] = $n3;
                $this->setCodeValue($mode3, 3, $n3);

                $this->pos += 4;
                break;
            case '02':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];
                // $pos3 = $this->code[$this->pos + 3];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                $n3 = $n1 * $n2;
                // $this->code[$pos3] = $n3;
                $this->setCodeValue($mode3, 3, $n3);

                $this->pos += 4;
                break;
            case '03':
                if (count($this->inputs) > 0) {
                    $input = array_shift($this->inputs);
                } else {
                    $input = false;
                }

                if ($input === false) {
                    $this->status = self::STATUS_WAIT;
                } else {
                    // if ($mode1 === '0') {
                    //     $pos1 = $this->code[$this->pos + 1];
                    //     $this->code[$pos1] = $input;
                    // } else {
                    //     $this->code[$this->pos + 1] = $input;
                    // }
                    $this->setCodeValue($mode1, 1, $input);

                    $this->pos += 2;
                }
                break;
            case '04':
                // if ($mode1 === '0') {
                //     $pos1 = $this->code[$this->pos + 1];
                //     $this->signal = $this->code[$pos1];
                // } else {
                //     $this->signal = $this->code[$this->pos + 1];
                // }
                $this->signal = $this->getCodeValue($mode1, 1);

                $this->pos += 2;
                $this->status = self::STATUS_WAIT;
                break;
            case '05':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                if ($n1 != '0') {
                    $jump = $n2 - $this->pos;
                } else {
                    $jump = 3;
                }
                $this->pos += $jump;
                break;
            case '06':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                if ($n1 == '0') {
                    $jump = $n2 - $this->pos;
                } else {
                    $jump = 3;
                }
                $this->pos += $jump;
                break;
            case '07':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];
                // $pos3 = $this->code[$this->pos + 3];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                if ($n1 < $n2) {
                    // $this->code[$pos3] = 1;
                    $this->setCodeValue($mode3, 3, 1);
                } else {
                    // $this->code[$pos3] = 0;
                    $this->setCodeValue($mode3, 3, 0);
                }
                $this->pos += 4;
                break;
            case '08':
                // $pos1 = $this->code[$this->pos + 1];
                // $pos2 = $this->code[$this->pos + 2];
                // $pos3 = $this->code[$this->pos + 3];

                // $n1 = $mode1 === '0' ? $this->code[$pos1] : $pos1;
                // $n2 = $mode2 === '0' ? $this->code[$pos2] : $pos2;
                $n1 = $this->getCodeValue($mode1, 1);
                $n2 = $this->getCodeValue($mode2, 2);

                if ($n1 == $n2) {
                    // $this->code[$pos3] = 1;
                    $this->setCodeValue($mode3, 3, 1);
                } else {
                    // $this->code[$pos3] = 0;
                    $this->setCodeValue($mode3, 3, 0);
                }
                $this->pos += 4;
                break;
            case '09':
                $n1 = $this->getCodeValue($mode1, 1);
                $this->relative_base += $n1;
                $this->pos += 2;
                break;
        }

        return [$op, $this->status];
    }

    public function getSignal()
    {
        return $this->signal;
    }

    public function getStatus()
    {
        return $this->status;
    }

    public function getCode()
    {
        return implode(',', $this->code);
    }
}

File: machines/intcode.py
from typing import List


class Intcode:
    STATUS_STOP = 0
    STATUS_RUNNING = 1
    STATUS_WAIT = 2
    STATUS_HALT = 3

    __code: List[str]
    __status: int
    __signal: str
    __pos: int
    __inputs: List[str]
    __relative_base: int

    def __init__(self, intcode):
        self.__code = intcode.split(",")
        self.__status = self.STATUS_RUNNING
        self.__signal = None
        self.__pos = 0
        self.__inputs = []
        self.__relative_base = 0

    def set_code(self, pos: int, code: str):
        self.__code[pos] = code
        return True

    def set_inputs(self, inputs):
        if type(inputs) is list:
            self.__inputs += inputs
        else:
            self.__inputs += [inputs]
        self.__status = self.STATUS_RUNNING
        return True

    def __get_code_value(self, mode, pos_index):
        pos_value = self.__code[self.__pos + pos_index]
        return_value = False
        if int(mode) == 1:
            return_value = pos_value
        elif int(mode) == 2:
            idx = self.__relative_base + int(pos_value)
            return_value = self.__code[idx] if idx < len(self.__code) else "0"
        else:
            idx = int(pos_value)
            return_value = self.__code[idx] if idx < len(self.__code) else "0"
        return return_value

    def __set_code_value(self, mode, pos_index, value):
        value = str(value)
        pos_value = self.__code[self.__pos + pos_index]
        final_index = None
        if int(mode) == 2:
            final_index = self.__relative_base + int(pos_value)
        else:
            final_index = int(pos_value)

        if final_index >= len(self.__code):
            self.__code += [None] * (final_index - len(self.__code) + 1)

        self.__code[final_index] = value
        return True

    def calc(self):
        inst = f"{int(self.__code[self.__pos]):05d}"

        op = inst[-2:]
        mode1 = inst[2:3]
        mode2 = inst[1:2]
        mode3 = inst[0:1]

        jump = False

        if op == "99":
            self.__status = self.STATUS_HALT
        elif op == "01":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            n3 = int(n1) + int(n2)
            self.__set_code_value(mode3, 3, n3)

            self.__pos += 4
        elif op == "02":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            n3 = int(n1) * int(n2)
            self.__set_code_value(mode3, 3, n3)

            self.__pos += 4
        elif op == "03":
            if len(self.__inputs) > 0:
                input = self.__inputs.pop(0)
            else:
                input = False

            if input is False:
                self.__status = self.STATUS_WAIT
            else:
                self.__set_code_value(mode1, 1, input)
                self.__pos += 2
        elif op == "04":
            self.__signal = self.__get_code_value(mode1, 1)

            self.__pos += 2
            self.__status = self.STATUS_WAIT
        elif op == "05":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            if n1 != "0":
                jump = int(n2) - self.__pos
            else:
                jump = 3

            self.__pos += jump
        elif op == "06":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            if n1 == "0":
                jump = int(n2) - self.__pos
            else:
                jump = 3

            self.__pos += jump
        elif op == "07":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            if int(n1) < int(n2):
                self.__set_code_value(mode3, 3, 1)
            else:
                self.__set_code_value(mode3, 3, 0)
            self.__pos += 4
        elif op == "08":
            n1 = self.__get_code_value(mode1, 1)
            n2 = self.__get_code_value(mode2, 2)

            if int(n1) == int(n2):
                self.__set_code_value(mode3, 3, 1)
            else:
                self.__set_code_value(mode3, 3, 0)
            self.__pos += 4
        elif op == "09":
            n1 = self.__get_code_value(mode1, 1)
            self.__relative_base += int(n1)
            self.__pos += 2

        return [op, self.__status]

    def get_signal(self):
        return self.__signal

    def get_status(self):
        return self.__status

    def get_code(self):
        return ",".join(self.__code)
